### ISSUE TITLE
[4.2] Roundup - what to do with FormRequests & Validation

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/CRUD#3964]
We have a quite few PRs that are interconnected/alternatives. We need to take a look at _the big picture_ and have a clear unified position on how Inputs & FormRequests & Validation work inside Backpack.

Here's a round-up:
- [x] in https://github.com/Laravel-Backpack/CRUD/pull/3789 we're using adding the ability to edit the request inside FormRequest, and the resulting request will be processed by Backpack;
- [x] in https://github.com/Laravel-Backpack/CRUD/pull/3792 we're thinking about saving only what was validated by the request; I've just added an idea in the PR which would make that easy, but also a lot more;
- [ ] in https://github.com/Laravel-Backpack/CRUD/pull/3533 we're thinking of making it easy to add validation logic inside CrudController (have validation on Create/Update, but without a FormRequest);

In addition:
- [ ] in https://github.com/Laravel-Backpack/CRUD/pull/3928 we try to add another logic layer, before saving a field to the db; that way we can turn `Some title` to `Some Title` and maybe even include the file uploading logic right then-and-there, instead of having it inside a mutator on the Model;

The problem is... they ALL sound like good ideas to me! I see where each and every one of them could be useful 😅

-----

So here's what I think:

1. We MUST do https://github.com/Laravel-Backpack/CRUD/pull/3789 . This will enable devs to do the same thing as https://github.com/Laravel-Backpack/CRUD/pull/3928, but in the FormRequest.

2. We SHOULD polish and merge https://github.com/Laravel-Backpack/CRUD/pull/3792 . This will enable devs to do the same thing as https://github.com/Laravel-Backpack/CRUD/pull/3928, but in the CrudController.

3. We COULD polish and merge https://github.com/Laravel-Backpack/CRUD/pull/3533 , to allow inline field validation and manual array validation;

4. We COULD do https://github.com/Laravel-Backpack/CRUD/pull/3928, but we could also NOT do it, since the same thing can already be done both in the FormRequest and in the Controller (see 1 and 2 above). Though... none of those would be _as convenient_ as defining that logic on the field itself. And especially if we do (3)... it would make a lot of sense to define everything about a field in one place. So in total a field definition could be:
```php
CRUD::field('title')->type('text')->validationRules('required|min:5|max:255')->beforeSaving(function($value) {
    return Str::of($value)->title();
});
```

--- 

Bottom line... I think we should do all of them 😅 This helped me structure my logic around these issues. Nobody _has to_ read it, but if anybody does and has a different opinion, please let me know.

**Comment by pxpm** [comment]
Note: `1` does not invalidate `4` as user might not be using FormRequests and `4` runs AFTER the validation, technically `4` is more a `decorator` than a `validato`, and I expect people would continue to write `'name' => 'required|min:2'` in the FormRequest (if they use it) and `strtoupper()` after validation in `4`.

**Comment by tabacitu** [comment]
Update: we're going to do all of them. 

1&2 have already been merged.
For 3 take a look at https://github.com/Laravel-Backpack/CRUD/pull/4004
For 4 take a look at https://github.com/Laravel-Backpack/CRUD/pull/3928#issuecomment-990001381

This roundup has served its purpose, so I'm going to close it. Let's move the conversation to the issues above.
